text/0000-ide.md

```diff
@@ -0,0 +1,530 @@
+
+
+## Other shared functionality
+
+Other functionality, such as refactoring and reformatting will be provided by
+separate tools rather than the oracle. These should be sharable between IDE
+implementations. They are not covered in this RFC.
+
+
+# Motivation
+
+An IDE collects together many tools into a single piece of software. Some of
+these are entirely separate from the rest of the Rust eco-system (such as editor
+functionality), some will reuse existing tools in pretty much the same way they
+are already used (e.g., formatting code, which should straightforwardly use
+Rustfmt), and some will have totally new ways of using the compiler or other
+tools (e.g., code completion).
+
+Modern IDEs are large and complex pieces of software; creating a new one from
+scratch for Rust would be impractical. Therefore we need to work with existing
+IDEs (such as Eclipse, IntelliJ, and Visual Studio) to provide functionality.
+These IDEs provide excellent editor and project management support out of the
+box, but know nothing about the Rust language.
+
+An important aspect of IDE support is that response times must be extremely
+short. Users expect information as they type. Running normal compilation of an
+entire project is far too slow. Furthermore, as the user is typing, the program
+will not be a valid, complete Rust program.
+
+We expect that an IDE may have its own lexer and parser. This is necessary for
+the IDE to quickly give parse errors as the user types. Editors are free to rely
+on the compiler's parsing if they prefer (the compiler will do its own parsing
+in any case). Further information (name resolution, type information, etc.) will
+be provided by the compiler via the oracle.
+
+
+# Detailed design
+
+## Quick-check compilation
+
+(See also open questions, below).
+
+We run the quick-check compiler on a single crate. At some point after quick
+checking, dependent crates must be rebuilt. This is the responsibility of an
+external tool to manage (see below). Quick-check is driven by an IDE (or
+possibly by the oracle), not by Cargo.
+
+
+### Incremental and lazy compilation
+
+Incremental compilation is where, rather than re-compiling an entire crate, only
+code which is changed and its dependencies are re-compiled. See
+[RFC #1298](https://github.com/rust-lang/rfcs/pull/1298).
+
+Lazy compilation is where, rather than compiling an entire crate, we start by
+compiling a single function (or possibly some other unit of code), and re-
+compiling code which is depended on until we are done. Not all of a crate will
+be compiled in this fashion.
+
+These two compilation strategies are faster than the current compilation model
+(compile everything, every time). They are somewhat orthogonal - compilation can
+be either lazy or incremental without implying the other. The [current
+proposal](https://github.com/rust-lang/rfcs/pull/1298) for supporting
+incremental compilation involves some lazy compilation as an implementation
+detail.
+
```

Comment:
Seems like "incremental" here is use to describe push-driven in contrast to the lazy pull-driven compilation. Which is more efficient depends strongly on the use case, hence supporting both and combinations is probably useful. If the dependencies, changes (the pushes) and interests (the pulls) are managed explicitly, combining the strategies should be feasible. I would use the word "incremental" instead to describe all of these partial compilations.